### PR TITLE
feat: faster rfc3339 parsing

### PR DESCRIPTION
**Pull request by sby1ce**
PR for #1735 

Improving performance of `DateTime::parse_from_rfc3339` by avoiding usage of `Parsed` and unrolling some loops. 

My simple benchmark shows 2 times improvement (~100 ns to ~50 ns). 
Though this still doesn't reach `speedate`'s ~30 ns, this is an easy `win` in optimization. 

Parsing function now performs some edge case checks that were done by `Parsed` to pass existing tests in the crate. 

Draft PR to collect feedback and also I would like to touch on `parse_rfc3339_relaxed` which is used for `serde` deserialization the same way

### CODE REVIEW

File: src/format/parse.rs [hunk 1]
```diff
@@ -189,40 +189,114 @@ pub(crate) fn parse_rfc3339<'a>(parsed: &mut Parsed, mut s: &'a str) -> ParseRes
...
+            Some(c) if c.is_ascii_digit() => Ok(T::from(c - b'0')),
+            _ => Err(INVALID),
+        }
+    }
 
-    s = match s.as_bytes().first() {
-        Some(&b't' | &b'T' | &b' ') => &s[1..],
-        Some(_) => return Err(INVALID),
-        None => return Err(TOO_SHORT),
+    let bytes: &[u8] = s.as_bytes();
```
Comment (djc) [inline src/format/parse.rs:200]: Please get rid off all unnecessary type annotations.

File: src/format/parse.rs [hunk 2]
```diff
@@ -189,40 +189,114 @@ pub(crate) fn parse_rfc3339<'a>(parsed: &mut Parsed, mut s: &'a str) -> ParseRes
...
-    parsed.set_month(try_consume!(scan::number(s, 2, 2)))?;
-    s = scan::char(s, b'-')?;
-    parsed.set_day(try_consume!(scan::number(s, 2, 2)))?;
+    #[inline]
+    fn get_digit<T: From<u8>>(bytes: &[u8], index: usize) -> ParseResult<T> {
+        match bytes.get(index) {
+            Some(c) if c.is_ascii_digit() => Ok(T::from(c - b'0')),
+            _ => Err(INVALID),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (sby1ce) [reply src/format/parse.rs:198]: Ok, I'll do that but the reason for this function is nested is that it's a small utility function only used in this place, meaning outside it may get separated. 
It also carries semantic meaning for the reader of the code to see that while technically the function performs a bounds check, the `None` case is actually optimized away by the compiler as the function declaration is right next to the `bytes.len()` check.

**Comment by sby1ce** [comment]
> Okay, let's start with a separate PR that adds a benchmark. The code style in this PR is not up to my standards and I'm not motivated to write up a zillion style comments and keep going back and forth about it.

Ok I'm up for anything. I'm not sure what kind of benchmark you're looking for though. 
There's already one
https://github.com/chronotope/chrono/blob/0061d86cb531c469272c22727c414a0eb9e55816/bench/benches/chrono.rs#L30-L37

**Comment by sby1ce** [comment]
> > My simple benchmark shows 2 times improvement (~100 ns to ~50 ns).
> > Though this still doesn't reach `speedate`'s ~30 ns, this is an easy `win` in optimization.
> 
> I mean this benchmark. Is that the one you were referring to?

Basically, yes. A benchmark simply calling a parser function on a string.

**Comment by djc** [comment]
See

- #1748
